File: algorithms/0230.kth-smallest-element-in-a-bst/solution.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} k
# @return {Integer}
def kth_smallest(root, k)
  @vals = []
  helper(root)
  @vals.sort[k - 1]
end

def helper(node)
  return if node.nil?

  @vals << node.val
  helper(node.left)
  helper(node.right)
end

File: algorithms/0230.kth-smallest-element-in-a-bst/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func kthSmallest(root *TreeNode, k int) int {
	vals := []int{}
	helper(root, &vals)

	if len(vals) > 0 {
		sort.Ints(vals)
		return vals[k-1]
	} else {
		return -1 // unreachable
	}
}

func helper(node *TreeNode, vals *[]int) {
	if node == nil {
		return
	}

	helper(node.Left, vals)
	helper(node.Right, vals)

	*vals = append(*vals, node.Val)
}
